hs_core/hydroshare/features.py

```diff
@@ -0,0 +1,434 @@
+        apps = {}
+        for v in Variable.objects.filter(timestamp__gte=fromdate, timestamp__lte=todate):
+            user = v.session.visitor.user
+            if user is not None and \
+               user.username != 'test' and user.username != 'demo':
+                user_id = user.username
+                if v.name == 'app_launch':
+                    value = v.get_value()
+                    # print("user:{} value:{} action:{}".format(user_id, value, v.name))
+                    m = expr.search(value)  # resource short id
+                    if m and m.group(1):
+                        resource_id = m.group(1)
+                        user_id = user.username
+                        if resource_id not in apps:
+                            apps[user_id] = set([resource_id])
+                        else:
+                            apps[user_id].add(resource_id)
+        return apps
+
+    @staticmethod
+    def user_favorites():
+        favs = {}
+        for u in User.objects.filter(is_active=True):
+            for r in u.ulabels.favorited_resources:
+                if u.username not in favs:
+                    favs[u.username] = set([r.short_id])
+                else:
+                    favs[u.username].add(r.short_id)
+        return favs
+
+    @staticmethod
+    def user_my_resources():
+        mine = {}
+        for u in User.objects.filter(is_active=True):
+            for r in u.ulabels.my_resources:
+                if u.username not in mine:
+                    mine[u.username] = set([r.short_id])
+                else:
+                    mine[u.username].add(r.short_id)
+        return mine
+
+    @staticmethod
+    def user_owned_groups():
+        groups = {}
+        for u in User.objects.filter(is_active=True):
+            for g in u.uaccess.get_groups_with_explicit_access(PrivilegeCodes.OWNER):
+                if u.username not in groups:
+                    groups[u.username] = set([g.name])
+                else:
+                    groups[u.username].add(g.name)
+        return groups
+
+    @staticmethod
+    def user_edited_groups():
+        groups = {}
+        for u in User.objects.filter(is_active=True):
+            for g in u.uaccess.get_groups_with_explicit_access(PrivilegeCodes.CHANGE):
+                if u.username not in groups:
+                    groups[u.username] = set([g.name])
+                else:
+                    groups[u.username].add(g.name)
+        return groups
+
+    @staticmethod
+    def user_viewed_groups():
+        groups = {}
+        for u in User.objects.filter(is_active=True):
+            for g in u.uaccess.get_groups_with_explicit_access(PrivilegeCodes.VIEW):
+                if u.username not in groups:
+                    groups[u.username] = set([g.name])
+                else:
+                    groups[u.username].add(g.name)
+        return groups
+
+    @staticmethod
+    def resources_editable_via_group(g):
```

Comment:
Same comment: give me a block top-level comment that says what g is. Perhaps the name "group" would be better for readability.